### PR TITLE
feat(email-recovery): storage + smtp_request setup flow

### PR DESCRIPTION
**Pull request by sea-snake**
## Summary

PR 5+6 of the email-recovery stack — the storage layer plus the canister-side surface for the **setup (binding) flow**. Stacks on #3841 (DoH fallback).

The recovery half (`prepare_delegation`, `recover@id.ai` dispatch, delegation issuance) is deferred to a follow-up PR — it needs a stable reverse `address → AnchorNumber` index for the verified-`From:` lookup, which is best landed as its own focused change.

### What this PR adds

**Storage:**
- `EmailRecoveryCredential` lives **directly on the anchor struct** as `Option<EmailRecoveryCredential>` via a new `#[n(5)]` field on `StorableAnchor`. minicbor is forward-compatible across optional-field additions, so old anchors deserialize cleanly with `None` — no schema migration.
- Heap-only **pending-challenge map** (10k capacity, 30-min TTL, oldest-first eviction). Heap rather than stable: entries expire in 30 minutes and losing them on canister upgrade just costs an in-flight wizard one retry, not worth the stable-memory churn.
- **ChaCha20Rng PRNG**, seeded once per canister lifetime via `raw_rand`. Per-call cost is one PRNG draw, not a management-canister round trip.

**Canister surface (Candid):**
- `email_recovery_credential_prepare_add(IdentityNumber, EmailRecoveryDnsInput) -> Result<EmailRecoveryChallenge, EmailRecoveryError>` — authenticated update.
- `email_recovery_status(text) -> EmailRecoveryStatus` — anonymous query (FE polls).
- `email_recovery_credential_remove(IdentityNumber, text) -> Result<(), EmailRecoveryError>` — authenticated update.
- `smtp_request(SmtpRequest) -> SmtpResponse` — open update; the gateway entry point.

**Two verification paths (FE is path-agnostic):**

The FE just submits `{ address, selector, dns_proof: Option<DnsProofBundle> }`. The canister picks the path:

| Path | When | DKIM key sourced from | Outcall at email time? |
|---|---|---|---|
| **DNSSEC** | `dns_proof` is `Some` | Validated at prepare time → cached on pending challenge | No |
| **DoH allowlist** | `dns_proof` is `None` | `crate::doh::fetch_txt` (3-of-5 quorum, heap-cached) | Yes (first email per domain per TTL) |
| **Neither** | No proof + domain not allowlisted | — | Rejects with `DomainNotSupported` / `DomainNotAllowlisted` |

The DNSSEC path validates the chain synchronously against `DnssecConfig.root_anchors` and confirms the bundle's leaf is at `<selector>._domainkey.<domain>`. Candid-friendly mirror types (`DnsProofBundle`, `SignedRRset`, `Rrsig`, `DelegationLink`) are defined in the interface crate with `From<>` conversions to the canister-internal verifier types.

**Verification pipeline (smtp_request):**
- Recipient dispatch on the **full** address (`register@id.ai`) — defence-in-depth against direct callers spoofing just the user-part.
- Pulls the canister-issued nonce from the `Subject:` header (case-insensitive prefix match, hex-suffix length-checked).
- Sources the DKIM key (cached DNSSEC TXT or DoH fetch).
- Optionally fetches `_dmarc.<domain>`; absence forces strict alignment in the verifier.
- Runs `crate::dmarc::verify_email` end-to-end (full DKIM + DMARC + `Subject` in `h=`).
- On success: writes the credential to the anchor and flips the pending challenge to `Succeeded`. Failures flip to `Failed(reason)`. The gateway always sees `Ok`.

### What this PR does NOT add (intentional, deferred to follow-ups)

- **Recovery flow** — `prepare_delegation`, `recover@id.ai` dispatch, delegation issuance. Needs a reverse address index in stable storage; cleaner as its own PR.
- **Frontend** — the wizard UX from design-doc §8.6 is its own stack of PRs.

## Test plan

- [x] 441 unit tests pass: PRNG, pending-challenge map (TTL, eviction, status flips), prepare_add (input validation, allowlist gate, domain rejection, happy path), remove (happy path, case-insensitive match, idempotency edge cases), smtp.rs nonce extraction.
- [x] 12 PocketIC integration tests:
  - Nonce shape, allowlist gate, authz, status lifecycle (Pending → Expired), TTL expiry, silent-drop paths (no nonce, unknown nonce, wrong recipient).
  - **DNSSEC path**: no-trust-anchors rejection; DNSSEC-takes-precedence-over-DoH-allowlist.
  - **Full DoH e2e**: prepare → in-test DKIM signer → smtp_request → mocked DoH outcalls → RegistrationSucceeded → credential_remove succeeds.
- [x] `cargo check --target wasm32-unknown-unknown` clean.
- [x] `cargo clippy --tests -- -D warnings` clean.
- [x] `cargo fmt --check` clean.
- [x] `.did` interface stays in sync (the existing `check_candid_interface_compatibility` test enforces this).
- [x] Happy-path DNSSEC e2e test (with a real `_domainkey` chain).
- [ ] Smoke test on local replica — deferred to FE wizard integration.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## PR Stack
| # | PR | Description | Status |
|---|---|---|---|
| 0 | [#3836](https://github.com/dfinity/internet-identity/pull/3836) | Design doc | Open |
| 1 | [#3838](https://github.com/dfinity/internet-identity/pull/3838) | DNSSEC verifier scaffold | Open |
| 2 | [#3839](https://github.com/dfinity/internet-identity/pull/3839) | DKIM verifier (RFC 6376) | Open |
| 3 | [#3840](https://github.com/dfinity/internet-identity/pull/3840) | DMARC alignment (RFC 7489) | Open |
| 4 | [#3841](https://github.com/dfinity/internet-identity/pull/3841) | DoH fallback | Open |
| 5+6 | [#3842](https://github.com/dfinity/internet-identity/pull/3842) | Setup flow (storage + smtp_request) | Open |
| 7 | [#3843](https://github.com/dfinity/internet-identity/pull/3843) | Recovery flow (delegation) | Open |
| 8 | [#3844](https://github.com/dfinity/internet-identity/pull/3844) | Frontend + feature flag | Open |
| 9 | [#3855](https://github.com/dfinity/internet-identity/pull/3855) | Deploy/upgrade scripts: dnssec_config + doh_config | Open |
| 10 | [#3857](https://github.com/dfinity/internet-identity/pull/3857) | Email-recovery UX overhaul | Open |

## Changes during review

- The on-anchor `email_recovery` field is now a `Vec` (`Option<Vec<...>>` on the storable form, matching the existing `passkey_credentials`/`recovery_keys` pattern). The canister API still enforces ≤1 entry — this is a data-model widening so future multi-credential support doesn't need another schema bump. `Option<Vec<T>>` means anchors written under the previous schema decode cleanly with `email_recovery: None` (verified empirically against the in-tree minicbor-derive version).

### CODE REVIEW

File: src/internet_identity/src/email_recovery/prepare.rs [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+        created_at_secs: now_secs,
+        status: PendingStatus::Pending,
+    };
+    insert_with_eviction(nonce.clone(), challenge, now_secs);
+
+    Ok(EmailRecoveryChallenge {
+        nonce,
+        mailbox: super::SETUP_MAILBOX.into(),
+        expires_at: now_secs.saturating_add(super::CHALLENGE_TTL_SECS),
+    })
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/email_recovery/prepare.rs:142]: Fixed in c7d12b1a. `expires_at` is now in nanoseconds since epoch — matching the rest of II `Timestamp` encoding.

File: src/internet_identity/src/email_recovery/prepare.rs [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+                    .any(|d| d.eq_ignore_ascii_case(&registered_domain))
+            })
+            .unwrap_or(false)
+    });
+    if !allowlisted {
+        // The FE doesn't need to distinguish "DNSSEC unavailable AND
+        // DoH not allowlisted" from "neither path applies" — both
+        // collapse to "we can't accept email from this domain."
+        return Err(EmailRecoveryError::DomainNotSupported(registered_domain));
+    }
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/email_recovery/prepare.rs]: Fixed in c7d12b1a. Now returns `DomainNotAllowlisted(domain)` for the DoH-path miss. `DomainNotSupported` is reserved for genuinely unsupported domain shapes (and for the "DNSSEC trust anchors not configured" case in 1daaacb2).

File: src/internet_identity/src/email_recovery/smtp.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+    let envelope = match request.envelope.as_ref() {
+        Some(e) => e,
+        None => return smtp_err(SMTP_ERR_SYNTAX_ERROR, "Missing envelope"),
+    };
+    let recipient_user = envelope.to.user.to_ascii_lowercase();
+    if recipient_user != SETUP_RECIPIENT_USER {
+        // Drop with Ok — we don't emit a per-recipient signal back
+        // to the gateway.
+        return SmtpResponse::Ok {};
+    }
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/email_recovery/smtp.rs:151]: Fixed in c7d12b1a. Recipient dispatch now matches on the full address (user + domain) via a new `recipient_matches` helper. New `SETUP_RECIPIENT_DOMAIN` constant.

File: src/internet_identity/src/email_recovery/smtp.rs [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+    // Fetch DKIM TXT and DMARC TXT (best-effort) via DoH. The DoH
+    // module's heap cache means repeat traffic for the same
+    // selector + domain is a microsecond cache hit.
+    let dkim_fqdn = format!("{}._domainkey.{}", snapshot.selector, domain);
+    let dmarc_fqdn = format!("_dmarc.{}", domain);
+
+    let dkim_bytes = crate::doh::fetch_txt(&dkim_fqdn, &domain)
+        .await
+        .map_err(|e| EmailRecoveryError::DohFetchFailed(format!("dkim: {e:?}")))?;
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/email_recovery/smtp.rs]: Fixed in c7d12b1a. New `map_doh_error` helper maps `DohError` variants to typed `EmailRecoveryError`s with proper semantics.

File: src/internet_identity_interface/src/internet_identity/types/email_recovery.rs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+//!   fresh session keypair.
+//!
+//! Both flows hinge on the canister verifying a DKIM-signed email
+//! whose DKIM key it trusts. The trust comes from one of two paths:
+//!
+//! 1. **DoH allowlist** — for major mailbox providers (Gmail, Outlook,
+//!    iCloud, …) the canister fetches the DKIM TXT via a 5-of-3 quorum
+//!    across independent DoH providers (`crate::doh::fetch_txt`).
+//!    Restricted to a deploy-arg allowlist. **This is the only path
+//!    supported in the initial cut.**
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity_interface/src/internet_identity/types/email_recovery.rs]: Fixed in c7d12b1a — corrected to 3-of-5.

File: src/internet_identity_interface/src/internet_identity/types/email_recovery.rs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    /// ChaCha20Rng PRNG seeded once via `raw_rand`.
+    pub nonce: String,
+    /// Where the user should send the email — `register@id.ai` for
+    /// setup, `recover@id.ai` for recovery. The canister identifies
+    /// the challenge from the `nonce` in the `Subject:`, not from
+    /// the recipient address.
+    pub mailbox: String,
+    /// Unix seconds. 30 minutes after issue.
+    pub expires_at: Timestamp,
+}
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity_interface/src/internet_identity/types/email_recovery.rs]: Fixed in c7d12b1a. Timestamp docs corrected to "nanoseconds since epoch". Nonce format corrected to "16 lowercase hex characters".

File: src/internet_identity/src/storage/storable/email_recovery_credential.rs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    /// [`EmailRecoveryCredential::address`].
+    #[n(0)]
+    pub address: String,
+    /// Unix-seconds.
+    #[n(1)]
+    pub created_at: Timestamp,
+    /// Unix-seconds — `None` until the first `smtp_request` actually
+    /// uses this credential to authorise a recovery.
+    #[n(2)]
+    pub last_used: Option<Timestamp>,
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/storage/storable/email_recovery_credential.rs]: Fixed in c7d12b1a. Both fields now correctly documented as nanoseconds.

File: src/internet_identity/src/email_recovery/remove.rs [hunk 8]
```diff
@@ -0,0 +1,136 @@
+//! `email_recovery_credential_remove` — detach the recovery email.
+//!
+//! Wrapped in `anchor_operation_with_authz_check` at the canister-
+//! method layer (`main.rs::email_recovery_api`), so by the time we
+//! get here the caller is verified, the anchor is loaded, and we
+//! just need to clear the field.
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply src/internet_identity/src/email_recovery/remove.rs]: Fixed in c7d12b1a. Module doc now accurately describes the inlined auth + write pattern.

**Comment by sea-snake** [comment]
Replaced with a new PR from an upstream branch (enables direct collaboration). Same content, new PR number.
